src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
+    Tuple,
+    Unit,
+}
+
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
 pub struct StructDef {
     /// Fields, not including ctor
     pub fields: Vec<StructField>,
     /// ID of the constructor. This is only used for tuple- or enum-like
     /// structs.
-    pub ctor_id: Option<NodeId>,
+    pub id: NodeId,
```

Comment:
@arielb1 
>It may be better to use the ctor id as the "variant id" of a tuple-like struct, but I didn't try that.

Isn't that what https://github.com/petrochenkov/rust/commit/5abb670d41d7ae37c0fb87d79994b64000c19e6f does? It merges "variant id" and "ctor id" and as a result we have one id in the outer `enum`/`struct` item with `enum`/`struct` type and another id inside of `StructDef` with `fn` type (also used as "variant id").